File: library/src/main/java/com/upstream/basemvvmimpl/presentation/utils/PreferenceManager.java
package com.upstream.basemvvmimpl.presentation.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.text.TextUtils;

public class PreferenceManager {

    private final String TAG = "PreferenceManager";

    public static class Builder {
        private boolean isShared = true;
        private String filename;
        private Context context;

        public Builder setContext(Context context) {
            this.context = context;
            return this;
        }

        public Builder setShared(boolean value) {
            this.isShared = value;
            return this;
        }

        public Builder setFilename(String filename) {
            this.filename = filename;
            return this;
        }

        public PreferenceManager build() throws IllegalArgumentException {

            if (context == null) throw new IllegalArgumentException("No context!");

            if (!isShared) {
                return new PreferenceManager(context);
            } else {
                if (TextUtils.isEmpty(filename))
                    throw new IllegalArgumentException("Filename cannot be empty!");
                else {
                    filename = context.getPackageName() + "." + filename;
                    return new PreferenceManager(context, filename);
                }
            }
        }
    }

    private SharedPreferences sharedPreferences;
    private SharedPreferences.Editor editor;

    private PreferenceManager(Context context, String preference_filename) throws IllegalArgumentException {
        sharedPreferences = context.getSharedPreferences(preference_filename, Context.MODE_PRIVATE);
    }

    private PreferenceManager(Context context) {
        sharedPreferences = android.preference.PreferenceManager.getDefaultSharedPreferences(context);
    }

    public boolean contains(String preference) {
        return sharedPreferences.contains(preference);
    }

    public void setOnSharedPreferenceChangeListener(SharedPreferences.OnSharedPreferenceChangeListener listener) {
        sharedPreferences.registerOnSharedPreferenceChangeListener(listener);
    }

    public void removeOnSharedPreferenceChangeListener(SharedPreferences.OnSharedPreferenceChangeListener listener) {
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(listener);
    }

    public void savePreference(String preference, int value) {
        editor = sharedPreferences.edit();
        editor.putInt(preference, value);
        editor.apply();
    }

    public void savePreference(String preference, String value) {
        editor = sharedPreferences.edit();
        editor.putString(preference, value);
        editor.apply();
    }

    public void savePreference(String preference, boolean value) {
        editor = sharedPreferences.edit();
        editor.putBoolean(preference, value);
        editor.apply();
    }

    public void savePreference(String preference, float value) {
        editor = sharedPreferences.edit();
        editor.putFloat(preference, value);
        editor.apply();
    }

    public void savePreference(String preference, long value) {
        editor = sharedPreferences.edit();
        editor.putLong(preference, value);
        editor.apply();
    }

    public String getStringPreference(String preference, String default_value) {
        return sharedPreferences.getString(preference, default_value);
    }

    public boolean getBoolPreference(String preference, boolean default_value) {
        return sharedPreferences.getBoolean(preference, default_value);
    }

    public int getIntPreference(String preference, int default_value) {
        return sharedPreferences.getInt(preference, default_value);
    }

    public long getLongPreference(String preference, long default_value) {
        return sharedPreferences.getLong(preference, default_value);
    }

    public Float getFloatPreference(String preference, float default_value) {
        return sharedPreferences.getFloat(preference, default_value);
    }
}

File: library/src/main/java/com/merseyside/mvvmcleanarch/utils/PreferenceManager.kt
package com.merseyside.mvvmcleanarch.utils

import android.content.Context
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener

class PreferenceManager private constructor(
    context: Context,
    preference_filename: String
) {

    class Builder(private val context: Context) {
        private var filename: String? = null

        fun setFilename(filename: String): Builder {
            this.filename = filename
            return this
        }

        @Throws(IllegalArgumentException::class)
        fun build(): PreferenceManager {
            return PreferenceManager(context,
                filename ?: "${context.packageName}.prefs"
            )
        }
    }

    private var sharedPreferences: SharedPreferences =
        context.getSharedPreferences(preference_filename, Context.MODE_PRIVATE)

    operator fun contains(preference: String?): Boolean {
        return sharedPreferences.contains(preference)
    }

    fun setOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener?) {
        sharedPreferences.registerOnSharedPreferenceChangeListener(listener)
    }

    fun removeOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener?) {
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(listener)
    }

    fun put(preference: String?, value: Int) {
        sharedPreferences.edit().apply {
            putInt(preference, value)
            apply()
        }
    }

    fun put(preference: String?, value: String?) {
        sharedPreferences.edit().apply {
            putString(preference, value)
            apply()
        }
    }

    fun put(preference: String?, value: Boolean) {
        sharedPreferences.edit().apply {
            putBoolean(preference, value)
            apply()
        }
    }

    fun put(preference: String?, value: Float) {
        sharedPreferences.edit().apply {
            putFloat(preference, value)
            apply()
        }
    }

    fun put(preference: String?, value: Long) {
        sharedPreferences.edit().apply {
            putLong(preference, value)
            apply()
        }
    }

    fun getString(
        preference: String,
        default_value: String
    ): String {
        return sharedPreferences.getString(preference, default_value)!!
    }

    fun getString(
        preference: String
    ): String? {
        return sharedPreferences.getString(preference, null)
    }

    fun getBool(
        preference: String,
        default_value: Boolean
    ): Boolean {
        return sharedPreferences.getBoolean(preference, default_value)
    }

    fun getInt(preference: String, default_value: Int): Int {
        return sharedPreferences.getInt(preference, default_value)
    }

    fun getLong(preference: String, default_value: Long): Long {
        return sharedPreferences.getLong(preference, default_value)
    }

    fun getFloat(preference: String, default_value: Float): Float {
        return sharedPreferences.getFloat(preference, default_value)
    }
}